### PR TITLE
Search jobs: improve tests

### PR DESCRIPTION
**Pull request by camdencheek**
This adds some unit tests for our search jobs just as a "these are doing what we think they are, right?" check.

### CODE REVIEW

File: internal/search/run/expression_job.go [hunk 1]
```diff
@@ -168,9 +173,12 @@ func (j *OrJob) Run(ctx context.Context, db database.DB, stream streaming.Sender
 	}
 
 	// Send results that were only seen by some of the sources
-	stream.Send(streaming.SearchEvent{
```
Comment (camdencheek) [inline internal/search/run/expression_job.go:171]: We would send an empty event if there were no unsent matches.

File: internal/search/run/expression_job.go [hunk 2]
```diff
@@ -86,6 +86,11 @@ func (a *AndJob) Name() string {
 // NewAndJob creates a job that will run each of its child jobs and stream
 // deduplicated matches that were streamed by at least one of the jobs.
 func NewOrJob(children ...Job) Job {
+	if len(children) == 0 {
+		return NewNoopJob()
+	} else if len(children) == 1 {
+		return children[0]
+	}
```
Comment (camdencheek) [inline internal/search/run/expression_job.go:93]: This minor optimization wasn't applied to `OrJob`, even though it was implemented elsewhere.

**Comment by camdencheek** [comment]
I have no idea why this just notified for, like, everything. To all the people here from Codenotify, feel free to ignore this PR.

File: internal/search/run/expression_job_test.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+				var subexpressions []Job
+				for j := 0; j < i; j++ {
+					subexpressions = append(subexpressions, mj)
+				}
+
+				j := NewAndJob(subexpressions...)
+
+				start := time.Now()
+				var eventTime time.Time
+				s := streaming.StreamFunc(func(e streaming.SearchEvent) {
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:48]:
```suggestion
				s := streaming.StreamFunc(func(_ streaming.SearchEvent) {
```

(nit to assert I actually read through the code 😝 )

File: internal/search/run/expression_job_test.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			t.Run(fmt.Sprintf("%d subexpressions", i), func(t *testing.T) {
+				subexpressions := []Job{noSender}
+				for j := 1; j < i; j++ {
+					subexpressions = append(subexpressions, sender)
+				}
+
+				j := NewOrJob(subexpressions...)
+
+				var eventTime time.Time
+				s := streaming.StreamFunc(func(e streaming.SearchEvent) {
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:154]:
```suggestion
				s := streaming.StreamFunc(func(_ streaming.SearchEvent) {
```

File: internal/search/run/expression_job_test.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+				var subexpressions []Job
+				for j := 0; j < i; j++ {
+					subexpressions = append(subexpressions, mj)
+				}
+
+				j := NewOrJob(subexpressions...)
+
+				start := time.Now()
+				var eventTime time.Time
+				s := streaming.StreamFunc(func(e streaming.SearchEvent) {
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:124]:
```suggestion
				s := streaming.StreamFunc(func(_ streaming.SearchEvent) {
```

File: internal/search/run/expression_job_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+			t.Run(fmt.Sprintf("%d subexpressions", i), func(t *testing.T) {
+				subexpressions := []Job{noSender}
+				for j := 1; j < i; j++ {
+					subexpressions = append(subexpressions, sender)
+				}
+
+				j := NewAndJob(subexpressions...)
+
+				eventSent := false
+				s := streaming.StreamFunc(func(e streaming.SearchEvent) {
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:82]:
```suggestion
				s := streaming.StreamFunc(func(_ streaming.SearchEvent) {
```

File: internal/search/run/expression_job_test.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+					eventTime = time.Now()
+				})
+				_, err := j.Run(context.Background(), nil, s)
+				require.NoError(t, err)
+				require.Less(t, eventTime.Sub(start), time.Millisecond)
+			})
+		}
+	})
+
+	t.Run("result not returned from all subexpressions is not streamed until all subexpressions return", func(t *testing.T) {
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:134]:
```suggestion
	t.Run("result not streamed until all subexpressions return the same result", func(t *testing.T) {
```

Or some other rephrase, my brain hurts reading this, I think the "result not returned from" is throwing me

File: internal/search/run/expression_job_test.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	})
+
+	t.Run("result returned from all subexpressions is streamed", func(t *testing.T) {
+		res := &result.RepoMatch{Name: "test", ID: 1}
+		mj := NewMockJob()
+		mj.RunFunc.SetDefaultHook(func(_ context.Context, _ database.DB, s streaming.Sender) (*search.Alert, error) {
+			s.Send(streaming.SearchEvent{Results: []result.Match{res}})
+			time.Sleep(10 * time.Millisecond)
+			return nil, nil
+		})
```
Comment (rvantonder) [inline internal/search/run/expression_job_test.go:35]: nit: there seems to be enough repetition and bulk that this setup could be factored out? I'm having some trouble grokking the core differences for each `t.Run`

**Review by rvantonder — APPROVED** [review]
Some nits but overall yay! Tests!

File: internal/search/run/expression_job_test.go (hunk 8 above)
Comment (camdencheek) [reply internal/search/run/expression_job_test.go]: Oh yeah. On second glance, there was a whole lot of copypasta that I could dedupe

**Comment by rvantonder** [comment]
stellar

**Pull request merged**